plugin/trino-redshift/src/main/java/io/trino/plugin/redshift/RedshiftUnloadSplitSource.java

```diff
@@ -0,0 +1,115 @@
+ * limitations under the License.
+ */
+package io.trino.plugin.redshift;
+
+import com.amazon.redshift.util.RedshiftException;
+import com.google.common.collect.ImmutableList;
+import io.airlift.log.Logger;
+import io.trino.plugin.jdbc.JdbcSplit;
+import io.trino.spi.connector.ConnectorSplit;
+import io.trino.spi.connector.ConnectorSplitSource;
+
+import java.sql.Connection;
+import java.sql.PreparedStatement;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.util.List;
+import java.util.Optional;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static java.util.Objects.requireNonNull;
+
+public class RedshiftUnloadSplitSource
+        implements ConnectorSplitSource
+{
+    private static final Logger log = Logger.get(RedshiftUnloadSplitSource.class);
+    public static final String RESULT_FILES_QUERY = "SELECT rtrim(path) FROM stl_unload_log WHERE query=pg_last_query_id() ORDER BY path";
+
+    private final Connection connection;
+    private final CompletableFuture<Boolean> resultSetFuture;
+    private final ConnectorSplitSource fallbackSplitSource;
+    private boolean isUseFallbackSplitSource;
+
+    private boolean finished;
+
+    public RedshiftUnloadSplitSource(ExecutorService executor, Connection connection, PreparedStatement statement, ConnectorSplitSource fallbackSplitSource)
+    {
+        requireNonNull(executor, "executor is null");
+        this.connection = requireNonNull(connection, "connection is null");
+        this.fallbackSplitSource = requireNonNull(fallbackSplitSource, "fallbackSplitSource is null");
+        requireNonNull(statement, "statement is null");
+        resultSetFuture = CompletableFuture.supplyAsync(() -> {
+            log.debug("Executing: %s", statement);
+            try {
+                return statement.execute();
+            }
+            catch (SQLException e) {
+                throw new RuntimeException(e);
+            }
+        }, executor);
+    }
+
+    @Override
+    public CompletableFuture<ConnectorSplitBatch> getNextBatch(int maxSize)
+    {
+        if (isUseFallbackSplitSource) {
+            return fallbackSplitSource.getNextBatch(maxSize);
+        }
+        try {
+            resultSetFuture.get();
+        }
+        catch (InterruptedException | ExecutionException e) {
+            if (e.getCause() != null && e.getCause().getCause() != null && e.getCause().getCause() instanceof RedshiftException) {
+                isUseFallbackSplitSource = true;
+                return fallbackSplitSource.getNextBatch(maxSize);
+            }
+            throw new RuntimeException(e);
+        }
+        finished = true;
+        List<ConnectorSplit> splits = getPaths().stream()
+                .map(path -> new JdbcSplit(Optional.of(path)))
```

Comment:
Do we need to explot `JdbcSplit` here ? Previously we had a HandleResolver which would allow us to map only a single split implementation for a connector but now things are a bit dynamic - So can we have a custom split implementation for the Unload approach and JDBC based approach ?

Reply:
Good point. Added `RedshiftUnloadSplit`